astropy/units/physical.py

```diff
@@ -321,7 +321,7 @@ def __init__(self, unit, physical_types):
     def __iter__(self):
         yield from self._physical_type_list
 
-    def __eq__(self, other):
+    def __eq__(self, other: object) -> bool:
```

Comment:
Shouldn't this be
```suggestion
    def __eq__(self, other: PhysicalType | str) -> bool:
```

Reply:
It's valid to do ``PhysicalType() == 4``, which will return False after trying `4.__req__()`.

Reply:
Unresolving this, since I think @eerovaher's comment is actually relevant: isn't typing supposed to tell what is needed to make *this* function succeed? With `PhysicalType|str`, that is clear, with `object` you get zero information.

I guess for these dunder methods, it depends on what you mean by "succeed" - I'd argue that returning `NotImplemented` is a form of not succeeding, since if the reverse also does it, you get a `TypeError`. More generally, that the reverse is called is irrelevant for thinking about success of this function.

Reply:
See https://discuss.python.org/t/make-type-hints-for-eq-of-primitives-less-strict/34240/10 for a discussion where people are advocating for making typing dunder methods less strict. The typing of most dunder methods is limited by inheritance from "object". Currently the correct annotation for equality is `__eq__(self, other: object) -> bool`. Not returning NotImplemented is problematic but baked into Python ATM. Accepting any input is correct since evaluating an equality to False is a valid output, even if it happens from a double NotImplemented behind-the-scenes.

Reply:
Having `other` been restricted only makes sense if the method errors or doesn't work on that input. `__eq__` et al work, returning NotImplemented / False.

Reply:
I understand the logic, but it is not a useful one to humans or much in the spirit of what `NotImplemented` means - I'd say we don't annotate these until the discussion has concluded.